src/pybind/mgr/volumes/fs_subvolume.py

```diff
@@ -0,0 +1,249 @@
+            except cephfs.ObjectNotFound:
+                self.fs.mkdir(subpath, mode)
+
+    def create_subvolume(self, subvolume_path, size=None, namespace_isolated=True, mode=0o755):
+        """
+        Set up metadata, pools and auth for a subvolume.
+
+        This function is idempotent.  It is safe to call this again
+        for an already-created subvolume, even if it is in use.
+
+        :param subvolume_path: SubvolumePath instance
+        :param size: In bytes, or None for no size limit
+        :param namespace_isolated: If true, use separate RADOS namespace for this subvolume
+        :return: None
+        """
+        path = self._subvolume_path(subvolume_path)
+        log.info("create_subvolume: {0}".format(path))
+
+        self._mkdir_p(path, mode)
+
+        if size is not None:
+            self.fs.setxattr(path, 'ceph.quota.max_bytes', to_bytes(size), 0)
+
+        # enforce security isolation, use separate namespace for this subvolume
+        if namespace_isolated:
+            namespace = "{0}{1}".format(self.pool_ns_prefix, subvolume_path.subvolume_id)
+            log.info("create_subvolume: {0}, using rados namespace {1} to isolate data.".format(subvolume_path, namespace))
+            self.fs.setxattr(path, 'ceph.dir.layout.pool_namespace',
+                             to_bytes(namespace), 0)
+        else:
+            # If subvolume's namespace layout is not set, then the subvolume's pool
+            # layout remains unset and will undesirably change with ancestor's
+            # pool layout changes.
+            pool_name = self._get_ancestor_xattr(path, "ceph.dir.layout.pool")
+            self.fs.setxattr(path, 'ceph.dir.layout.pool',
+                             to_bytes(pool_name), 0)
+
+    def delete_subvolume(self, subvolume_path):
+        """
+        Delete a subvolume in two parts. First, make a subvolume inaccessible to guests
+        by moving it to a trash folder. This is the fast part. Then, clear up the subvolume
+        and its contents, which is the slow part.
+
+        :param subvolume_path: Same identifier used in create_subvolume
+        :return: None
+        """
+
+        path = self._subvolume_path(subvolume_path)
+        log.info("delete_subvolume: {0}".format(path))
+
+        # Create the trash folder if it doesn't already exist
+        trash = os.path.join(self.subvolume_prefix, "_deleting")
+        self._mkdir_p(trash)
+
+        # We'll move it to here
+        trashed_subvolume = os.path.join(trash, subvolume_path.subvolume_id + str(uuid.uuid4()))
+
+        # Move the subvolume to the trash folder
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            log.warning("Trying to delete subvolume '{0}' but it's already gone".format(
+                path))
+            return
+        else:
+            self.fs.rename(path, trashed_subvolume)
+
+        # Purge the subvolume
+        self._purgedir(trashed_subvolume)
```

Comment:
rename + purge is what's done in currently in manila, and ceph-csi,
See,  https://github.com/ceph/ceph-csi/blob/csi-v1.0/pkg/cephfs/volume.go#L105

Since purge (rm -rf) might take a while, a prior rename to trash folder would immediately make the subvolume's contents inaccessible.